src/audio/asrc/asrc_ipc3.c

```diff
@@ -0,0 +1,1048 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2019-2022 Intel Corporation. All rights reserved.
+
+#include <sof/audio/asrc/asrc_farrow.h>
+#include <sof/audio/buffer.h>
+#include <sof/audio/component.h>
+#include <sof/audio/format.h>
+#include <sof/audio/pipeline.h>
+#include <sof/audio/ipc-config.h>
+#include <rtos/panic.h>
+#include <sof/ipc/msg.h>
+#include <rtos/alloc.h>
+#include <rtos/init.h>
+#include <sof/lib/memory.h>
+#include <sof/lib/uuid.h>
+#include <sof/math/numbers.h>
+#include <sof/trace/trace.h>
+#include <sof/common.h>
+#include <sof/compiler_attributes.h>
+#include <sof/list.h>
+#include <sof/platform.h>
+#include <rtos/string.h>
+#include <sof/ut.h>
+#include <ipc/control.h>
+#include <ipc/stream.h>
+#include <ipc/topology.h>
+#include <user/trace.h>
+#include <errno.h>
+#include <stddef.h>
+#include <stdint.h>
+
+/* Simple count value to prevent first delta timestamp
+ * from being input to low-pass filter.
+ */
+#define TS_STABLE_DIFF_COUNT	2
+
+/* Low pass filter coefficient for measured drift factor,
+ * The low pass function is y(n) = c1 * x(n) + c2 * y(n -1)
+ * coefficient c2 needs to be 1 - c1.
+ */
+#define COEF_C1		Q_CONVERT_FLOAT(0.01, 30)
```

Comment:
A lot of code looks generic and common. Why not keep asrc.c as the common module api client and split ipc3 and ipc4 specifics to separate source files. It's not sustainable to duplicate the code for every new IPC major version.

Reply:
It is also very possible that IPC3 operation is permanently broken due to other changes to SOF. Could consider erroring for IPC3 build and remove all specifics if we are not able to validate it.

Reply:
since we will convert to module interface finally, and it would be much different. so I plan to split it first. I just compile it by docker for IPC3 and did not see any error, did not verify it on IPC3 yet